### PR TITLE
add Backbone.Collection.clone()

### PR DESCRIPTION
**Pull request by ianfp**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://backbonejs.org/#Collection-clone
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/backbone/index.d.ts [hunk 1]
```diff
@@ -229,6 +229,7 @@ declare namespace Backbone {
          **/
         get(id: number|string|Model): TModel;
         has(key: number|string|Model): boolean;
+        clone(): Collection<TModel>;
```
Comment (DanielRosenwasser) [inline types/backbone/index.d.ts]: Would it make more sense to write this as

```ts
clone(): this;
```

?

**Comment by jaapz** [comment]
@ianfp why did you use `this` as a return type for `clone`? `Collection<Model>` makes more sense, right?

**Comment by ianfp** [comment]
@jaapz A return type of `this` in Typescript does not mean "this instance" -- it means "whatever subclass this is". See https://www.typescriptlang.org/docs/handbook/advanced-types.html#polymorphic-this-types

For example, if the return type were `Collection<TModel>` and I had a custom method like this:

```typescript
class MyCollection extends Collection<MyModel> {
   public someCustomMethod(): void {}
}
```
Then this fluent usage would trigger a type error.
```typescript
const coll = new MyCollection();
coll.clone().someCustomMethod();
```
The error message would be something like "class Collection does not have a method called someCustomMethod()".

**Comment by jaapz** [comment]
That's weird, because when I override clone like this:

```typescript
class MyCollection extends Collection<MyModel> {
    clone() {
        return new MyCollection(this.models, someClonedOptions);
    }
}
```

I get an error from TypeScript saying `MyCollection` is not allowed to be returned in this function. I can call super.clone() instead, but then it doesn't copy all options, so I have to use setters to also copy the options over to the new collection.

For reference, this is how backbone implements clone, as you can see it doesn't copy over all options:

```javascript
    clone: function() {
      return new this.constructor(this.models, {
        model: this.model,
        comparator: this.comparator
      });
    },
```

**Comment by ianfp** [comment]
@jaapz Ouch -- you are right. I just updated `@types/backbone` and got bit by my own "fix". :-(

I think this illustrates the problem:
```typescript
class MyCollection extends Collection<MyModel> {
    clone(): this {
        return new MyCollection(...);
    }
}

class MySpecificCollection extends MyCollection {
    // nothing is overridden
}

const original = new MySpecificCollection();
const copy = mc.clone();
```
What is the type of `copy`?

It's `MyCollection`, of course, but according to the return type of `clone()`, it should be `MySpecificCollection`. [This Typescript issue](https://github.com/Microsoft/TypeScript/issues/5863) suggests that there's no solution to this problem yet, but the following workaround seems to do the trick:

```typescript
 class MyCollection extends Collection<MyModel> {
    clone(): this {
        return new MyCollection(...) as this;
    }
}
```

At least, that's what I think is going on.

**Comment by jaapz** [comment]
Ah yes, nice work figuring out the problem, thanks. Maybe we can just add a comment to the `clone` type definition telling people how this should be done if you're not using `super`, and linking to the TypeScript issue? I can't seem to think of a better way to actually fix this.

**Pull request merged**